dev/change-scala-version.sh

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) 2023, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+set -e
+
+VALID_VERSIONS=( 2.12 2.13 )
+
+usage() {
+  echo "Usage: $(basename $0) [-h|--help] <version>
+where :
+  -h| --help Display this help text
+  valid version values : ${VALID_VERSIONS[*]}
+" 1>&2
+  exit 1
+}
+
+if [[ ($# -ne 1) || ( $1 == "--help") ||  $1 == "-h" ]]; then
+  usage
+fi
+
+TO_VERSION=$1
+
+check_scala_version() {
+  for i in ${VALID_VERSIONS[*]}; do [ $i = "$1" ] && return 0; done
+  echo "Invalid Scala version: $1. Valid versions: ${VALID_VERSIONS[*]}" 1>&2
+  exit 1
+}
+
+check_scala_version "$TO_VERSION"
+
+if [ $TO_VERSION = "2.13" ]; then
+  FROM_VERSION="2.12"
+else
+  FROM_VERSION="2.13"
+fi
+
+sed_i() {
+  sed -e "$1" "$2" > "$2.tmp" && mv "$2.tmp" "$2"
+}
+
+BASEDIR=$(dirname $0)/..
+for f in $(find "$BASEDIR" -name 'pom.xml' -not -path '*target*'); do
```

Comment:
one question: are we plan to override pom file in-place? 
this may change the git status (use unstaged file to build) of local repo which seems weird.

is this possible to be achieved by using 
1. override versions in profiles, or
2. put the temp file to .gitignore and use `maven --file` to specify the temp one if needed

Reply:
I don't think we can completely override the version in profiles.  One of the ways the Scala version appears is in the artifact ID, and last I checked you cannot make that a variable.  It's OK to put this as a variable in your dependencies, but not in the primary artifact ID of the pom.  I think this is one of the reasons Spark went with a script to change between versions rather than putting it in the pom as they've done with almost all other aspects of the build.

Reply:
I see, thanks. Then I would prefer the way to specify mvn running with generated new pomFile,
and also put the file name pattern in .gitignore to avoid changing git status if possible

Reply:
Another option is to remove the Scala version from artifact names or leave it unchanged. Then make sure that the correct pom.xml w.r.t to 2.12 vs 2.13 artifactId is generated to the  build `target` dir, prior to deploy-file plugin is run.


Just thinking out-loud:

If it was not for the shuffle manager, I think we could get away with a single jar (the size would still be dominated by libcudf(jni) jar entries).  This is because all other entry points like Driver/ExecutorPluin and ResourceDiscoveryPlugin are in Java. So we could refactor ShimLoader to only use Java and reflection to set up parallel worlds with the Scala version in the path after dynamically detecting both Spark and Scala versions `jar:file:/home/spark/rapids-4-spark-23.10.0.jar!/2.13/spark3xx-common/`

Reply:
I'm curious, only Spark 3.2.0 and above support Scala 2.13 and we only plan to support it with Spark 3.3.0 and above. Would this above strategy still work?

Reply:
if you are asking about the pom generation, customizing build/deploy for specific versions is easy since we already have buildver-specific profiles anyways.

Reply:
Will address these concerns in https://github.com/NVIDIA/spark-rapids/issues/9331